AnswerTest.testPartitionService

_master_ (commit a31d0df545758b469189b9d57152fd97f2cd1c18)

Failed on nightly build : https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/311/testReport/com.hazelcast.test.starter.answer.test/AnswerTest/testPartitionService/

Stacktrace : 

```
java.lang.NoSuchFieldError: 
Field localReplica does not exist on object Partition [0]{
}
	at com.hazelcast.test.starter.ReflectionUtils.getFieldValueReflectively(ReflectionUtils.java:140)
	at com.hazelcast.test.starter.constructor.InternalPartitionImplConstructor.createNew0(InternalPartitionImplConstructor.java:47)
	at com.hazelcast.test.starter.constructor.AbstractStarterObjectConstructor.createNew(AbstractStarterObjectConstructor.java:45)
	at com.hazelcast.test.starter.HazelcastProxyFactory.construct(HazelcastProxyFactory.java:316)
	at com.hazelcast.test.starter.HazelcastProxyFactory.constructWithoutProxy(HazelcastProxyFactory.java:276)
	at com.hazelcast.test.starter.HazelcastProxyFactory.proxyObjectForStarter(HazelcastProxyFactory.java:191)
	at com.hazelcast.test.starter.HazelcastProxyFactory.toArray(HazelcastProxyFactory.java:340)
	at com.hazelcast.test.starter.HazelcastProxyFactory.proxyObjectForStarter(HazelcastProxyFactory.java:177)
	at com.hazelcast.test.starter.answer.AbstractAnswer.invoke(AbstractAnswer.java:266)
	at com.hazelcast.test.starter.answer.AbstractAnswer.invoke(AbstractAnswer.java:222)
	at com.hazelcast.test.starter.answer.AbstractAnswer.invoke(AbstractAnswer.java:171)
	at com.hazelcast.test.starter.answer.PartitionServiceAnswer.answer(PartitionServiceAnswer.java:41)
	at com.hazelcast.test.starter.answer.AbstractAnswer.answer(AbstractAnswer.java:98)
	at org.mockito.internal.handler.MockHandlerImpl.handle(MockHandlerImpl.java:104)
	at org.mockito.internal.handler.NullResultGuardian.handle(NullResultGuardian.java:29)
	at org.mockito.internal.handler.InvocationNotifierHandler.handle(InvocationNotifierHandler.java:35)
	at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor.doIntercept(MockMethodInterceptor.java:63)
	at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor.doIntercept(MockMethodInterceptor.java:49)
	at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor$DispatcherDefaultingToRealMethod.interceptAbstract(MockMethodInterceptor.java:128)
	at com.hazelcast.internal.partition.InternalPartitionService$MockitoMock$1213416615.getPartitions(Unknown Source)
	at com.hazelcast.test.starter.answer.test.AnswerTest.testPartitionService(AnswerTest.java:189)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

*master* (commit 33082885b9b021e53bfabd4bf1589f21dec97d52)

This has started to fail constantly after update to 3.13-SNAPSHOT. Easily reproducible locally.

Stacktrace:
```
[INFO] Running com.hazelcast.test.starter.answer.test.AnswerTest
[ERROR] Tests run: 11, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 41.654 s <<< FAILURE! - in com.hazelcast.test.starter.answer.test.AnswerTest
[ERROR] testPartitionService(com.hazelcast.test.starter.answer.test.AnswerTest)  Time elapsed: 2.466 s  <<< ERROR!
java.lang.IllegalArgumentException: argument type mismatch
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at com.hazelcast.test.starter.constructor.InternalPartitionImplConstructor.createNew0(InternalPartitionImplConstructor.java:59)
	at com.hazelcast.test.starter.constructor.AbstractStarterObjectConstructor.createNew(AbstractStarterObjectConstructor.java:45)
	at com.hazelcast.test.starter.HazelcastProxyFactory.construct(HazelcastProxyFactory.java:316)
	at com.hazelcast.test.starter.HazelcastProxyFactory.constructWithoutProxy(HazelcastProxyFactory.java:276)
	at com.hazelcast.test.starter.HazelcastProxyFactory.proxyObjectForStarter(HazelcastProxyFactory.java:191)
	at com.hazelcast.test.starter.HazelcastProxyFactory.toArray(HazelcastProxyFactory.java:340)
	at com.hazelcast.test.starter.HazelcastProxyFactory.proxyObjectForStarter(HazelcastProxyFactory.java:177)
	at com.hazelcast.test.starter.answer.AbstractAnswer.invoke(AbstractAnswer.java:266)
	at com.hazelcast.test.starter.answer.AbstractAnswer.invoke(AbstractAnswer.java:222)
	at com.hazelcast.test.starter.answer.AbstractAnswer.invoke(AbstractAnswer.java:171)
	at com.hazelcast.test.starter.answer.PartitionServiceAnswer.answer(PartitionServiceAnswer.java:41)
	at com.hazelcast.test.starter.answer.AbstractAnswer.answer(AbstractAnswer.java:98)
	at org.mockito.internal.handler.MockHandlerImpl.handle(MockHandlerImpl.java:104)
	at org.mockito.internal.handler.NullResultGuardian.handle(NullResultGuardian.java:29)
	at org.mockito.internal.handler.InvocationNotifierHandler.handle(InvocationNotifierHandler.java:35)
	at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor.doIntercept(MockMethodInterceptor.java:63)
	at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor.doIntercept(MockMethodInterceptor.java:49)
	at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor$DispatcherDefaultingToRealMethod.interceptAbstract(MockMethodInterceptor.java:128)
	at com.hazelcast.internal.partition.InternalPartitionService$MockitoMock$1051967996.getPartitions(Unknown Source)
	at com.hazelcast.test.starter.answer.test.AnswerTest.testPartitionService(AnswerTest.java:189)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:745)
```

--- a/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/InternalPartitionImplConstructor.java
+++ b/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/InternalPartitionImplConstructor.java
@@ -21,9 +21,7 @@ import com.hazelcast.test.starter.HazelcastStarterConstructor;
 import java.lang.reflect.Array;
 import java.lang.reflect.Constructor;
 
-import static com.hazelcast.internal.partition.PartitionReplica.UNKNOWN_UID;
 import static com.hazelcast.test.starter.HazelcastProxyFactory.proxyArgumentsIfNeeded;
-import static com.hazelcast.test.starter.ReflectionUtils.getAllFieldsByName;
 import static com.hazelcast.test.starter.ReflectionUtils.getFieldValueReflectively;
 
 @HazelcastStarterConstructor(classNames = {"com.hazelcast.internal.partition.impl.InternalPartitionImpl"})
@@ -38,67 +36,20 @@ public class InternalPartitionImplConstructor extends AbstractStarterObjectConst
         ClassLoader classloader = targetClass.getClassLoader();
         Class<?> partitionListenerClass = classloader.loadClass("com.hazelcast.internal.partition.PartitionListener");
         Class<?> replicaClass = classloader.loadClass("com.hazelcast.internal.partition.PartitionReplica");
-        Class<?> addressClass = classloader.loadClass("com.hazelcast.nio.Address");
         Class<?> replicaArrayClass = Array.newInstance(replicaClass, 0).getClass();
         // obtain reference to constructor InternalPartitionImpl(int partitionId, PartitionListener listener,
         //                                                       PartitionReplica localReplica, PartitionReplica[] replicas)
         Constructor<?> constructor = targetClass.getDeclaredConstructor(Integer.TYPE, partitionListenerClass, replicaClass,
                 replicaArrayClass);
 
-        // obtain reference to constructor PartitionReplica(Address address, String uuid)
-        Constructor<?> replicaConstructor = replicaClass.getConstructor(addressClass, String.class);
-
         Integer partitionId = (Integer) getFieldValueReflectively(delegate, "partitionId");
         Object partitionListener = getFieldValueReflectively(delegate, "partitionListener");
-        Object localReplica = getLocalReplica(delegate, classloader, replicaConstructor);
-        Object replicas = getReplicas(delegate, classloader, replicaConstructor);
+        Object localReplica = getFieldValueReflectively(delegate, "localReplica");
+        Object replicas = getFieldValueReflectively(delegate, "replicas");
 
-        Object[] args = new Object[]{partitionId, partitionListener, localReplica, replicas};
+        Object[] args = new Object[] {partitionId, partitionListener, localReplica, replicas};
         Object[] proxiedArgs = proxyArgumentsIfNeeded(args, classloader);
 
         return constructor.newInstance(proxiedArgs);
-    }
-
-    private Object getReplicas(Object delegate,
-                               ClassLoader classloader,
-                               Constructor<?> replicaConstructor) throws Exception {
-        // RU_COMPAT_3_11
-        boolean is311Instance = is311Instance(delegate);
-        if (is311Instance) {
-            Object[] addresses = (Object[]) getFieldValueReflectively(delegate, "addresses");
-            Object[] replicas = (Object[]) Array.newInstance(replicaConstructor.getDeclaringClass(), addresses.length);
-            for (int i = 0; i < addresses.length; i++) {
-                replicas[i] = toPartitionReplica(addresses[i], replicaConstructor, classloader);
-            }
-            return replicas;
-        } else {
-            return getFieldValueReflectively(delegate, "replicas");
-        }
-    }
-
-    private Object getLocalReplica(Object delegate, ClassLoader classloader, Constructor<?> replicaConstructor) throws Exception {
-        // RU_COMPAT_3_11
-        boolean is311Instance = is311Instance(delegate);
-        if (is311Instance) {
-            Object thisAddress = getFieldValueReflectively(delegate, "thisAddress");
-            return toPartitionReplica(thisAddress, replicaConstructor, classloader);
-        } else {
-            return getFieldValueReflectively(delegate, "localReplica");
-        }
-    }
-
-    private static boolean is311Instance(Object delegate) {
-        return getAllFieldsByName(delegate.getClass()).containsKey("thisAddress");
-    }
-
-    private Object toPartitionReplica(Object address,
-                                      Constructor<?> replicaConstructor,
-                                      ClassLoader targetClassloader) throws Exception {
-        if (address == null) {
-            return null;
-        }
-        Object[] args = {address, UNKNOWN_UID};
-        Object[] proxiedArgs = proxyArgumentsIfNeeded(args, targetClassloader);
-        return replicaConstructor.newInstance(proxiedArgs);
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/MemberImplConstructor.java
+++ b/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/MemberImplConstructor.java
@@ -44,38 +44,21 @@ public class MemberImplConstructor extends AbstractStarterObjectConstructor {
         Boolean liteMember = (Boolean) getFieldValueReflectively(delegate, "liteMember");
 
         ClassLoader targetClassloader = targetClass.getClassLoader();
-        Class<?> addressClass = targetClassloader.loadClass("com.hazelcast.nio.Address");
         Class<?> memberVersionClass = targetClassloader.loadClass("com.hazelcast.version.MemberVersion");
 
-        try {
-            // RU_COMPAT_3_11
-            // obtain reference to constructor MemberImpl(Address address, MemberVersion version, boolean localMember,
-            //                                            String uuid, Map<String, Object> attributes, boolean liteMember)
-            Constructor<?> constructor = targetClass
-                    .getDeclaredConstructor(addressClass, memberVersionClass, Boolean.TYPE, String.class, Map.class, Boolean.TYPE);
-            Object[] args = new Object[]{address, memberVersion, localMember, uuid, attributes, liteMember};
-
-            Object[] proxiedArgs = proxyArgumentsIfNeeded(args, targetClassloader);
-            return constructor.newInstance(proxiedArgs);
-        } catch (NoSuchMethodException e) {
-            // Since 3.12, constructor args are
-            // (Map<EndpointQualifier, Address> addresses, MemberVersion version, boolean localMember,
-            // String uuid, Map<String, Object> attributes, boolean liteMember, int memberListJoinVersion,
-            // HazelcastInstanceImpl instance)
-            Class<?> hzImplClass = targetClassloader.loadClass("com.hazelcast.instance.HazelcastInstanceImpl");
-            Constructor<?> constructor = targetClass
-                    .getDeclaredConstructor(Map.class, memberVersionClass, Boolean.TYPE, String.class, Map.class, Boolean.TYPE,
-                            Integer.TYPE, hzImplClass);
-            constructor.setAccessible(true);
-            Class<?> endpointQualifierClass =
-                    targetClassloader.loadClass("com.hazelcast.instance.EndpointQualifier");
-            Object memberEndpointQualifer = endpointQualifierClass.getDeclaredField("MEMBER").get(null);
-            Map addressMap = new HashMap();
-            addressMap.put(memberEndpointQualifer, address);
-            Object[] args = new Object[] {addressMap, memberVersion, localMember, uuid, attributes, liteMember,
-                                          MemberImpl.NA_MEMBER_LIST_JOIN_VERSION, null};
-            return constructor.newInstance(proxyArgumentsIfNeeded(args, targetClassloader));
-        }
+        Class<?> hzImplClass = targetClassloader.loadClass("com.hazelcast.instance.HazelcastInstanceImpl");
+        Constructor<?> constructor = targetClass
+                .getDeclaredConstructor(Map.class, memberVersionClass, Boolean.TYPE, String.class, Map.class, Boolean.TYPE,
+                        Integer.TYPE, hzImplClass);
+        constructor.setAccessible(true);
+        Class<?> endpointQualifierClass =
+                targetClassloader.loadClass("com.hazelcast.instance.EndpointQualifier");
+        Object memberEndpointQualifer = endpointQualifierClass.getDeclaredField("MEMBER").get(null);
+        Map addressMap = new HashMap();
+        addressMap.put(memberEndpointQualifer, address);
+        Object[] args = new Object[] {addressMap, memberVersion, localMember, uuid, attributes, liteMember,
+                                      MemberImpl.NA_MEMBER_LIST_JOIN_VERSION, null};
+        return constructor.newInstance(proxyArgumentsIfNeeded(args, targetClassloader));
     }
 
     private static Object getMemberVersion(Object delegate) throws Exception {

--- a/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/PartitionReplicaConstructor.java
+++ b/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/PartitionReplicaConstructor.java
@@ -1,0 +1,49 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.test.starter.constructor;
+
+import com.hazelcast.test.starter.HazelcastStarterConstructor;
+
+import java.lang.reflect.Constructor;
+
+import static com.hazelcast.test.starter.HazelcastProxyFactory.proxyArgumentsIfNeeded;
+import static com.hazelcast.test.starter.ReflectionUtils.getFieldValueReflectively;
+
+@HazelcastStarterConstructor(classNames = {"com.hazelcast.internal.partition.PartitionReplica"})
+public class PartitionReplicaConstructor extends AbstractStarterObjectConstructor {
+
+    public PartitionReplicaConstructor(Class<?> targetClass) {
+        super(targetClass);
+    }
+
+    @Override
+    Object createNew0(Object delegate)
+            throws Exception {
+        ClassLoader classloader = targetClass.getClassLoader();
+        Class<?> replicaClass = classloader.loadClass("com.hazelcast.internal.partition.PartitionReplica");
+        Class<?> addressClass = classloader.loadClass("com.hazelcast.nio.Address");
+        Constructor<?> constructor = targetClass.getDeclaredConstructor(addressClass, String.class);
+
+        Object address = getFieldValueReflectively(delegate, "address");
+        Object uuid = getFieldValueReflectively(delegate, "uuid");
+
+        Object[] args = new Object[] {address, uuid};
+        Object[] proxiedArgs = proxyArgumentsIfNeeded(args, classloader);
+
+        return constructor.newInstance(proxiedArgs);
+    }
+}

--- a/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/test/PartitionReplicaConstructorTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/test/starter/constructor/test/PartitionReplicaConstructorTest.java
@@ -1,0 +1,55 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.test.starter.constructor.test;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.internal.partition.InternalPartition;
+import com.hazelcast.internal.partition.InternalPartitionService;
+import com.hazelcast.internal.partition.PartitionReplica;
+import com.hazelcast.internal.partition.impl.InternalPartitionImpl;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.test.starter.constructor.PartitionReplicaConstructor;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class PartitionReplicaConstructorTest extends HazelcastTestSupport {
+
+    @Test
+    public void testConstructor() {
+        HazelcastInstance hz = createHazelcastInstance();
+        warmUpPartitions(hz);
+
+        InternalPartitionService partitionService = getNode(hz).getPartitionService();
+        InternalPartition[] partitions = partitionService.getInternalPartitions();
+        InternalPartitionImpl partition = (InternalPartitionImpl) partitions[0];
+        PartitionReplica replica = partition.getReplica(0);
+
+        PartitionReplicaConstructor constructor = new PartitionReplicaConstructor(PartitionReplica.class);
+        PartitionReplica clonedReplica = (PartitionReplica) constructor.createNew(replica);
+
+        assertEquals(replica.address(), clonedReplica.address());
+        assertEquals(replica.uuid(), clonedReplica.uuid());
+    }
+}

